README.md
![Laravel Thumbnails](https://github.com/pratiksh404/laravel-thumbnails/blob/master/img/laravel-thumbnail.png)

[![Stars](https://img.shields.io/github/stars/pratiksh404/laravel-thumbnails)](https://github.com/pratiksh404/laravel-thumbnails/stargazers) [![Latest Stable Version](https://poser.pugx.org/drh2so4/thumbnail/v)](//packagist.org/packages/drh2so4/thumbnail) [![StyleCI](https://github.styleci.io/repos/270181879/shield?branch=master)](https://github.styleci.io/repos/270181879?branch=master) [![Build Status](https://scrutinizer-ci.com/g/pratiksh404/laravel-thumbnails/badges/build.png?b=master)](https://scrutinizer-ci.com/g/pratiksh404/laravel-thumbnails/build-status/master) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/pratiksh404/laravel-thumbnails/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/pratiksh404/laravel-thumbnails/?branch=master) [![Total Downloads](https://poser.pugx.org/drh2so4/thumbnail/downloads)](//packagist.org/packages/drh2so4/thumbnail) [![License](https://poser.pugx.org/drh2so4/thumbnail/license)](//packagist.org/packages/drh2so4/thumbnail) [![Laravel News](https://img.shields.io/badge/Featured-Laravel%20News-blue)](https://github.com/pratiksh404)

## Laravel Thumbnail Generator

Package for uploading the image and saving that image along with it's thumbnail.

## What does it do ?

- Uploads Image
- Make its thumbnail i.e low quality, resized version of its parent image

## Why use thumbnails ?

The small file size of thumbnails makes it possible for website designers to offer visitors a lot of content immediately without increasing the loading time of the page.
Also why use full glory of that image if you just have to crunched it up to tiny space... Use thumbnail.

### Installation

Run Composer Require Command

```sh
$ composer require drh2so4/thumbnail
```

Use thumbnail trait to your model

```sh
<?php

namespace App;

use drh2so4\Thumbnail\Traits\thumbnail;
use Illuminate\Database\Eloquent\Model;

class Image extends Model
{
    use Thumbnail;

    protected $fillable = ['image'];
}

```

This model initially consists of two methods :-

- makeThumbnail
- thumbnail

### Usages

Package utilizes it's trait method, let us guide you to use that

## makeThumbnail

This method is responsible for actually uploading the image and making its thumbnail.
It takes one parameter i.e image fieldname (default = "image").

```sh
    public function store(Request $request)
    {
        $image = Image::create($this->validateData());
        $image->makeThumbnail('image'); //This handles uploading image and storing it's thumbnail
        return redirect('/imageUpload');
    }
```

Same can go with update method.

## thumbnail

Well, we created our thumbnail but how to use it, let me guide through that
When we uploaded image with name suppose "batman".
Image is upload with name batman-current_time_instant i.e (batman-1521549.jpg).

What about thumbnail...
well thumbnail uses it's parent image name followed by -size
i.e batman-1521549-medium-jpg, batman-1521549-small.jpg

## How to make thumbnail ?

There are the options you can have for making thumbnails :-

- Default Option
- Universal Custom Thumbnails
- Specfic Custom Thumbnails

### Default Option

you can just call the following and the packages will handle the rest

```sh

    $image->makeThumbnail('image'); //Here the first parameter is image attribute name saved in db

```

Note : if the attribute dedicated for storing image is named 'image' you don't have to pass image attribute name jusr use \$image->makeThumbnail();

### Universal Custom Thumbnails

here you should mention the thumbnails that you want to be applied on every case.
when you publish thumbnail.php config file you will find 'thumbnails' property where you can mention your custom thumbnails

```sh
    /*
    |--------------------------------------------------------------------------
    | Custom Thumbnail Creation
    |--------------------------------------------------------------------------
    | Uncomment to create...
    */

    /*     "thumbnails" => [
        [
            "thumbnail-name" => "medium",
            "thumbnail-width" => 800,
            "thumbnail-height" => 600,
            "thumbnail-quality" => 60
        ],
        [
            "thumbnail-name" => "small",
            "thumbnail-width" => 400,
            "thumbnail-height" => 300,
            "thumbnail-quality" => 30
        ]
    ] */
```

Note: This will override default option

### Specfic Custom Thumbnails

Suppose you have applied Universal Custom Thumbnails but need to have changes for specific image field then you can pass array of custom requirements :

```sh
        $thumbnails = [
            'storage' => 'customs/embed',
            'width' => '600',
            'height' => '400',
            'quality' => '70',
            'thumbnails' => [
                [
                    'thumbnail-name' => 'customSmall',
                    'thumbnail-width' => '300',
                    'thumbnail-height' => '200',
                    'thumbnail-quality' => '50'
                ]
            ]
        ];
        $image->makeThumbnail('image', $thumbnails);
```

### How about multiple image uploads

If you are performing multiple image upload at once pass image key to thumbnail array.
here \$img is one of the image in chunk of images passed

```sh
// Controller Store Method
public function store(Request $request)
{
    $image = Image::create(['images'=>$request->images]);
    foreach($request->images as $img)
    {
        $this->multipleImageUpload($image,$img);
    }
}
// Multiple Image Upload
    private function multipleImageUpload($image, $img)
    {
        $multiple = [
            'storage' => 'custom_test/folder/another_folder/image',
            'width' => '600',
            'height' => '400',
            'quality' => '70',
            'image' => $img,
            'thumbnails' => [
                [
                    'thumbnail-name' => 'small',
                    'thumbnail-width' => '300',
                    'thumbnail-height' => '200',
                    'thumbnail-quality' => '50'
                ]
            ]
        ];
        $image->makeThumbnail('image', $multiple);
    }
```

## How to use thumbnail ?

Just call as following

```sh
// Here the first parameter 'image' is the name of sttribute that is saved in db for image

    @foreach ($images as $image)
        <img src="{{asset($image->thumbnail('image','small'))}}"> // For small thumbnail
    <img src="{{asset($image->thumbnail('image','medium'))}}"> // For medium thumbnail
    @endforeach
```

if you are using custom thumbnail configured from config file just call as follows

```sh
// Here the first parameter 'image' is the name of sttribute that is saved in db for image
// Second parameter is the name of thumbnail that you gave in 'thumbnail-name' in the config file on custom thumbnail field called 'thumbnails'
    @foreach ($images as $image)
        <img src="{{asset($image->thumbnail('image','small'))}}"> // For small thumbnail
    <img src="{{asset($image->thumbnail('image','medium'))}}"> // For medium thumbnail
    @endforeach
```

Notice that parameter of function thumbnail is string same as value given for "thumbnail-name" in config file.

Thumbnail's image property is predefined but if you wish to change that publish it's config file thumbnail.php

```sh
php artisan vendor:publish --tag=thumbnail-config
```

### Image Property

You can obtaing the detail image property by using method imageDetail($image,$size)

```sh
// imageDetail method takes two parameter
// Model image attribute name (Mandatory)
// If you want thumbnail property pass its size (optional)

$image->imageDetail('image'); // Parent Image Detail Property
$image->imageDetail('image','small') // Small thumbnail related to parent image
```

### What Image Property/Detail Gives

## Default Thumbnail Image Properties

| Property  | Return Type | Description                           | Example                                  |
| --------- | ----------- | ------------------------------------- | ---------------------------------------- |
| image     | string      | Image path stored in DB               | \$image->imageDetail('image')->image     |
| name      | string      | Image Stored Name (without extension) | \$image->imageDetail('image')->name      |
| fullname  | string      | Image Stored Name (with extension)    | \$image->imageDetail('image')->fullname  |
| extension | string      | Image Extension Name                  | \$image->imageDetail('image')->extension |
| path      | string      | Image Storage Path                    | \$image->imageDetail('image')->path      |
| directory | string      | Image Stored Directory                | \$image->imageDetail('image')->directory |
| location  | string      | Image Full Location Path              | \$image->imageDetail('image')->location  |
| property  | array       | Image Property array                  | \$image->imageDetail('image')->property  |

Image Property array (\$image->imageDetail('image')->property)

| Property        | Return Type | Description                                        | Example                                                  |
| --------------- | ----------- | -------------------------------------------------- | -------------------------------------------------------- |
| real_name       | string      | Image Real Name (without timestamp and size label) | \$image->imageDetail('image')->property->name            |
| size            | integer     | Image Storage Size                                 | \$image->imageDetail('image')->property->fullname        |
| directory       | string      | Image Stored Directory                             | \$image->imageDetail('image')->property->directory       |
| location        | string      | Image Full Location Path                           | \$image->imageDetail('image')->property->location        |
| has_thumbnail   | boolean     | Image's Thumbnail Check                            | \$image->imageDetail('image')->property->has_thumbnail   |
| thumbnail_count | integer     | Image Thumbnail Count                              | \$image->imageDetail('image')->property->thumbnail_count |
| thumbnails      | array       | Return all thumbnail Detail                        | \$image->imageDetail('image')->property->thumbnails      |

Image Thumbnail Property
| Property | Return Type | Description  
| --------------- | ----------- | --------------------------------------------------
| image | string | Thumbnail Name  
| real_name | string | Thumbnail Real Name (without timestamp and size label)
| size | integer | Thumbnail Storage Size  
| created_date | Carbon | Thumbnail Created Date  
| path | string | Thumbnail Storage Path  
| directory | string | Thumbnail Stored Directory  
| location | string | Thumbnail Full Location Path

### Check if image has thumbnail

```sh
$image->hasThumbnail('image'); // Check for any availabe thumbnail
$image->hasThumbnail('image','small'); // Second paremater is thumbnail size check
```

### Obtain Image Thumbnail Count

```sh
$image->thumbnailCount('image');
```

### Hard Delete Image with Thumbnails

```sh
$image->imageDetail('image'); // First parameter is db attribute name for image
```

### Upload Only Umage

```sh
$solo_image = [
                'storage' => 'solo',
                'width' => '600',
                'height' => '300',
                'quality' => '70',
            ];
            $image->uploadImage('image', $solo_image); // Second Parameter is not necessary if default settings is to be applied
            // OR
            $image->uploadImage('image'); // Image Upload with default setting
```

Our config file looks like follows :-

```sh
<?php

return [


    /*
    |--------------------------------------------------------------------------
    |  Thumbnail Feature
    |--------------------------------------------------------------------------
    |
    | This option defines whether to use Package's Thumbnail feature or not
    | Default option is true
    |
    */
    'thumbnail' => true,

    /*
    |--------------------------------------------------------------------------
    | Thumbnail Qualities
    |--------------------------------------------------------------------------
    |
    | These options are default post image and its thumbnail quality
    |
    |
    */

    'image_quality' => 80,
    'medium_thumbnail_quality' => 60,
    'small_thumbnail_quality' => 30,

    /*
    |--------------------------------------------------------------------------
    | Default Image Fit Size
    |--------------------------------------------------------------------------
    |
    | These options is default post imahe height and width fit size
    |
    |
    */

    'img_width' => 1000,
    'img_height' => 800,

    'medium_thumbnail_width' => 800,
    'medium_thumbnail_height' => 600,

    'small_thumbnail_width' => 400,
    'small_thumbnail_height' => 300,

    /*
    |--------------------------------------------------------------------------
    | Image and Thumbnail Storage Path
    |--------------------------------------------------------------------------
    |
    | Define your default image storage location along with its thumbnail
    |
    |
    */

    "storage_path" => "uploads",

    /*
    |--------------------------------------------------------------------------
    | Custom Thumbnail Creation
    |--------------------------------------------------------------------------
    | Uncomment to create...
    */

    /*     "thumbnails" => [
        [
            "thumbnail-name" => "medium",
            "thumbnail-width" => 800,
            "thumbnail-height" => 600,
            "thumbnail-quality" => 60
        ],
        [
            "thumbnail-name" => "small",
            "thumbnail-width" => 400,
            "thumbnail-height" => 300,
            "thumbnail-quality" => 30
        ],
        ] */
];


```

Feel free to change the values

## Default Thumbnail Image Properties

| Thumbnail        | Width | Height | Quality |
| ---------------- | ----- | ------ | ------- |
| Uploaded Image   | 1000  | 800    | 80      |
| Medium Thumbnail | 800   | 600    | 60      |
| Small Thumbnail  | 400   | 300    | 30      |

![Laravel Thumbnails](https://github.com/pratiksh404/laravel-thumbnails/blob/master/img/thumbnails.png)

### Todos

- Error Handling
- Image Caching
- Maintainabilty

## Package Used

- http://image.intervention.io/

## License

MIT

**DOCTYPE NEPAL || DR.H2SO4**


CHANGELOG.md
# Changelog

All notable changes to `laravel-thumbnails` will be documented in this file

## 1.0.0

- makeThumbnail and thumbnail method Introduction

## 1.0.1

- Inline Custom Thumbnail Configuration Feature

## 1.0.2

- Multiple Image Upload Feature

## 1.0.3

- Image Caching Feature

## 1.0.4

- Image Detail Feature
  Changes: thumbnail_storage removed


config/thumbnail.php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    |  Thumbnail Feature
    |--------------------------------------------------------------------------
    |
    | This option defines whether to use Package's Thumbnail feature or not
    | Default option is true
    |
    */
    'thumbnail' => true,

    /*
    |--------------------------------------------------------------------------
    | Thumbnail Qualities
    |--------------------------------------------------------------------------
    |
    | These options are default post image and its thumbnail quality
    |
    |
    */

    'image_quality'            => 80,
    'medium_thumbnail_quality' => 60,
    'small_thumbnail_quality'  => 30,

    /*
    |--------------------------------------------------------------------------
    | Default Image Fit Size
    |--------------------------------------------------------------------------
    |
    | These options is default post imahe height and width fit size
    |
    |
    */

    'img_width'  => 1000,
    'img_height' => 800,

    'medium_thumbnail_width'  => 800,
    'medium_thumbnail_height' => 600,

    'small_thumbnail_width'  => 400,
    'small_thumbnail_height' => 300,

    /*
    |--------------------------------------------------------------------------
    | Image and Thumbnail Storage Path
    |--------------------------------------------------------------------------
    |
    | Define your default image storage location along with its thumbnail
    |
    |
    */

    'storage_path' => 'uploads',

    /*
    |--------------------------------------------------------------------------
    | Image Cached Time (Minutes)
    |--------------------------------------------------------------------------
    |
    | Option for setting image cached time
    |
    |
    */

    'image_cached_time' => 10,

    /*
    |--------------------------------------------------------------------------
    | Custom Thumbnail Creation
    |--------------------------------------------------------------------------
    | Uncomment to create...
    */

    /*     "thumbnails" => [
        [
            "thumbnail-name" => "medium",
            "thumbnail-width" => 800,
            "thumbnail-height" => 600,
            "thumbnail-quality" => 60
        ],
        [
            "thumbnail-name" => "small",
            "thumbnail-width" => 400,
            "thumbnail-height" => 300,
            "thumbnail-quality" => 30
        ],
        ] */
];


src/Traits/Thumbnail.php
<?php

namespace drh2so4\Thumbnail\Traits;

use Carbon\Carbon;
use Illuminate\Support\Facades\File;
use Intervention\Image\Facades\Image as Image;

trait Thumbnail
{
    public function makeThumbnail($fieldname = 'image', $custom = [])
    {
        if (! empty(request()->$fieldname) || $custom['image']) {
            /* ------------------------------------------------------------------- */

            $image_file = $custom['image'] ?? request()->file($fieldname); // Retriving Image File
            $extension = $this->image_info($image_file)->extension; //Retriving Image extension
            $imageStoreNameOnly = $this->image_info($image_file)->imageStoreNameOnly; //Making Image Store name

            /* ------------------------------------------------------------------- */

            /* ----------------------------------------Parent Image Upload----------------------------------------- */
            $this->uploadImage($fieldname, $custom); // Upload Parent Image
            /* --------------------------------------------------------------------------------------------- */
            if (config('thumbnail.thumbnail', true)) {
                $thumbnails = false;
                $thumbnails = $custom['thumbnails'] ?? config('thumbnail.thumbnails') ?? false; // Grab Thumbnails
                $storage = $custom['storage'] ?? config('thumbnail.storage_path', 'uploads') ?? false; // Grab Storage Info
                if ($thumbnails) {
                    /* -----------------------------------------Custom Thumbnails------------------------------------------------- */
                    $this->makeCustomThumbnails($image_file, $imageStoreNameOnly, $extension, $storage, $thumbnails);
                /* -------------------------------------------------------------------------------------------------- */
                } else {
                    /* ---------------------------------------Default Thumbnails--------------------------------------- */
                    $this->makeDefaultThumbnails($image_file, $extension, $imageStoreNameOnly);
                    /* ------------------------------------------------------------------------------------------------ */
                }
            }
        }
    }

    // Make Image
    private function makeImg($image_file, $name, $location, $width, $height, $quality)
    {
        $image = $image_file->storeAs($location, $name, 'public'); // Thumbnail Storage Information
        $img = Image::cache(function ($cached_img) use ($image_file, $width, $height) {
            return $cached_img->make($image_file->getRealPath())->fit($width, $height);
        }, config('thumbnail.image_cached_time', 10), true); //Storing Thumbnail
        $img->save(public_path('storage/'.$image), $quality); //Storing Thumbnail
    }

    // Make Custom Thumbnail
    private function makeCustomThumbnails($image_file, $imageStoreNameOnly, $extension, $storage, $thumbnails)
    {
        foreach ($thumbnails as $thumbnail) {
            $customthumbnail = $imageStoreNameOnly.'-'.str_replace('-', '', $thumbnail['thumbnail-name']).'.'.$extension; // Making Thumbnail Name
            $this->makeImg(
                $image_file,
                $customthumbnail,
                $storage,
                (int) $thumbnail['thumbnail-width'],
                (int) $thumbnail['thumbnail-height'],
                (int) $thumbnail['thumbnail-quality']
            );
        }
    }

    // Make Default Thumbnail
    private function makeDefaultThumbnails($image_file, $extension, $imageStoreNameOnly)
    {
        /* --------------------- Thumbnail Info---------------------------------------- */
        //small thumbnail name
        $smallthumbnail = $imageStoreNameOnly.'-small'.'.'.$extension; // Making Thumbnail Name

        //medium thumbnail name
        $mediumthumbnail = $imageStoreNameOnly.'-medium'.'.'.$extension; // Making Thumbnail Name

        // Medium Thumbnail
        $this->makeImg(
            $image_file,
            $mediumthumbnail,
            config('thumbnail.storage_path', 'uploads'),
            config('thumbnail.medium_thumbnail_width', 800),
            config('thumbnail.medium_thumbnail_height', 600),
            config('thumbnail.medium_thumbnail_quality', 60)
        );

        // Small Thumbnail
        $this->makeImg(
            $image_file,
            $smallthumbnail,
            config('thumbnail.storage_path', 'uploads'),
            config('thumbnail.small_thumbnail_width', 800),
            config('thumbnail.small_thumbnail_height', 600),
            config('thumbnail.small_thumbnail_quality', 60)
        );

        /* ------------------------------------------------------------------------------------- */
    }

    /* Image Upload Process Info */
    private function image_info($image_file)
    {
        $filenamewithextension = $image_file->getClientOriginalName(); //Retriving Full Image Name
        $raw_filename = pathinfo($filenamewithextension, PATHINFO_FILENAME); //Retriving Image Raw Filename only
        $filename = str_replace(['-', ' '], '', $raw_filename); // Retrive Filename
        $extension = $image_file->getClientOriginalExtension(); //Retriving Image extension
        $imageStoreNameOnly = $filename.'-'.time(); //Making Image Store name
        $imageStoreName = $filename.'-'.time().'.'.$extension; //Making Image Store name

        $image_info['filenamewithextension'] = $filenamewithextension;
        $image_info['raw_filename'] = $raw_filename;
        $image_info['filename'] = $filename;
        $image_info['extension'] = $extension;
        $image_info['imageStoreNameOnly'] = $imageStoreNameOnly;
        $image_info['imageStoreName'] = $imageStoreName;

        return json_decode(json_encode($image_info));
    }

    // Upload Parent Image
    public function uploadImage($fieldname = 'image', $custom = [])
    {
        $image_file = $custom['image'] ?? request()->file($fieldname); // Retriving Image File
        $img = $custom['image'] ?? request()->$fieldname;
        $imageStoreName = $this->image_info($image_file)->imageStoreName;
        $this->update([
            $fieldname => $img->storeAs($custom['storage'] ?? config('thumbnail.storage_path', 'uploads'), $imageStoreName, 'public'), // Storing Parent Image
        ]);

        $image = Image::cache(function ($cached_img) use ($image_file, $custom) {
            return $cached_img->make($image_file->getRealPath())->fit($custom['width'] ?? config('thumbnail.img_width', 1000), $custom['height'] ?? config('thumbnail.img_height', 800)); //Parent Image Interventing
        }, config('thumbnail.image_cached_time', 10), true);
        $image->save(public_path('storage/'.$this->$fieldname), $custom['quality'] ?? config('thumbnail.image_quality', 80)); // Parent Image Locating Save
    }

    // Thumbnail Path
    public function thumbnail($fieldname = 'image', $size = null)
    {
        return $this->imageDetail($fieldname, $size)->path;
    }

    /* Checking Image Existance */
    private function imageExists($image)
    {
        return file_exists($image->getRealPath());
    }

    // Checking Image's Thumbnail Existance
    public function hasThumbnail($fieldname = 'image', $size = null)
    {
        return $this->imageDetail($fieldname, $size)->property->has_thumbnail;
    }

    // Thumbnail Count
    public function thumbnailCount($fieldname = 'image', $size = null)
    {
        return $this->hasThumbnail($fieldname, $size) ? $this->imageDetail($fieldname, $size)->property->thumbnail_count : 0;
    }

    /* Image Details */
    public function imageDetail($fieldname = 'image', $size = null)
    {
        $image = $this->$fieldname;
        $path = explode('/', $image);
        $extension = \File::extension($image);
        $name = basename($image, '.'.$extension);
        $image_fullname = isset($size) ? $name.'-'.(string) $size.'.'.$extension : $name.'.'.$extension;
        array_pop($path);
        $location = implode('/', $path);
        $path = 'storage/'.$location.'/'.$image_fullname;
        $image_files = File::files(public_path('storage/'.$location));
        $images_property = $this->imageProperty($image_files, $name);
        $image_detail = [
            'image'     => $image,
            'name'      => $name,
            'fullname'  => $image_fullname,
            'extension' => $extension,
            'path'      => $path,
            'directory' => public_path('storage/'.$location),
            'location'  => public_path('storage/'.$image),
            'property'  => $images_property,
        ];

        return json_decode(json_encode($image_detail));
    }

    // Image Property
    private function imageProperty($image_files, $parent_name)
    {
        $images_property = [];
        $thumbnails_property = [];
        $thumbnail_count = 0;
        foreach ($image_files as $image) {
            $image_partition = explode('-', basename($image));
            $parent_thumbnail_name = $image_partition[0].'-'.$image_partition[1];
            if ($parent_name == $parent_thumbnail_name) {
                $thumbnail_count++;
                $thumbnail_exists = $this->imageExists($image);
                if (isset($image_partition[2])) {
                    $thumbnails_property['image'] = $image->getFilename() ?? null;
                    $thumbnails_property['real_name'] = $image_partition[0];
                    $thumbnails_property['size'] = $image->getSize();
                    $thumbnails_property['created_date'] = isset($image_partition[1]) ? Carbon::createFromFormat('Y/m/d H:i:s', date('Y/m/d H:i:s', (int) $image_partition[1])) : null;
                    $thumbnails_property['directory'] = $image->getPath();
                    $thumbnails_property['location'] = $image->getRealPath();
                    $images_property['has_thumbnail'] = $thumbnail_exists || $this->imageExists($image);
                    $images_property['thumbnail_count'] = $thumbnail_count;
                    $thumbnails[] = $thumbnails_property;
                    $images_property['thumbnails'] = $thumbnails;
                }
            } elseif ($image->getFileNameWithoutExtension() == $parent_name) {
                $images_property['has_thumbnail'] = ($thumbnail_exists ?? false);
                $images_property['real_name'] = $image_partition[0];
                $images_property['size'] = $image->getSize();
                $images_property['directory'] = $image->getPath();
                $images_property['location'] = $image->getRealPath();
            } else {
                false;
            }
        }

        return $images_property;
    }

    // Hard Delete
    public function hardDelete($fieldname = 'image'): void
    {
        if (File::exists($this->imageDetail($fieldname)->location)) {
            if ($this->imageDetail($fieldname)->property->has_thumbnail) {
                foreach ($this->imageDetail($fieldname)->property->thumbnails as $thumbnail) {
                    File::exists($thumbnail->location) ? File::delete($thumbnail->location) : '';
                }
            }
            File::exists($this->imageDetail($fieldname)->location) ? File::delete($this->imageDetail($fieldname)->location) : false;
        }
        $this->delete();
    }
}
